clang/test/CodeGenSYCL/sycl-device.cpp

```diff
@@ -3,6 +3,78 @@
 
 int bar(int b);
 
+int bar10(int a) { return a + 10; }
+int bar20(int a) { return a + 20; }
+
+class A {
+public:
+  // CHECK-DAG: define linkonce_odr spir_func void @_ZN1A3fooEv
+  // CHECK-DAG: define spir_func i32 @_Z5bar20i
+  __attribute__((sycl_device)) void foo() { bar20(10); }
+
+  // CHECK-DAG: define linkonce_odr spir_func void @_ZN1AC1Ev
+  __attribute__((sycl_device))
+  A() { bar10(10); }
+  // CHECK-DAG: define spir_func i32 @_Z5bar10i
+  // CHECK-DAG: define linkonce_odr spir_func void @_ZN1AD1Ev
+  __attribute__((sycl_device)) ~A() {}
+
+  template <typename T>
+  __attribute__((sycl_device)) void AFoo(T t) {}
+
+  // Templates are emitted when they are instantiated
+  // CHECK-DAG: define linkonce_odr spir_func void @_ZN1A4AFooIiEEvT_
+  template <>
+  __attribute__((sycl_device)) void AFoo<int>(int t) {}
+};
+
+template <typename T>
+struct B {
+  T data;
+  B(T _data) : data(_data) {}
+
+  __attribute__((sycl_device)) void BFoo(T t) {}
+};
+
+template <>
+struct B<int> {
+  int data;
+  B(int _data) : data(_data) {}
+
+  // CHECK-DAG: define linkonce_odr spir_func void @_ZN1BIiE4BFooEi
+  __attribute__((sycl_device)) void BFoo(int t) {}
+};
+
+struct Base {
+  // CHECK-DAG: define linkonce_odr spir_func void @_ZN4Base12BaseWithAttrEv
+  __attribute__((sycl_device)) virtual void BaseWithAttr() { int a = 10; }
+  virtual void BaseWithoutAttr() { int b = 20; }
```

Comment:
Could you add a test case with calling a member function w/o `sycl_device` attribute from the annotated member function, please?

Reply:
Sure, added on lines 30-34